tq/ssh.go

```diff
@@ -0,0 +1,390 @@
+	if err != nil {
+		return err
+	}
+	status, _, text, err := conn.ReadStatusWithArguments()
+	if err != nil {
+		return err
+	}
+	if status < 200 || status > 299 {
+		if len(text) > 0 {
+			return fmt.Errorf("got status %d when verifying upload OID %s: %s", status, t.Oid, text[0])
+		}
+		return fmt.Errorf("got status %d when verifying upload OID %s", status, t.Oid)
+	}
+	return nil
+}
+
+func (a *SSHAdapter) doUpload(t *Transfer, conn *ssh.PktlineConnection, f *os.File, cb ProgressCallback) (int, []string, []string, error) {
+	args := a.argumentsForTransfer(t, "upload")
+
+	// Ensure progress callbacks made while uploading
+	// Wrap callback to give name context
+	ccb := func(totalSize int64, readSoFar int64, readSinceLast int) error {
+		if cb != nil {
+			return cb(t.Name, totalSize, readSoFar, readSinceLast)
+		}
+		return nil
+	}
+
+	cbr := tools.NewFileBodyWithCallback(f, t.Size, ccb)
+
+	conn.Lock()
+	defer conn.Unlock()
+	defer cbr.Close()
+	err := conn.SendMessageWithData(fmt.Sprintf("put-object %s", t.Oid), args, cbr)
+	if err != nil {
+		return 0, nil, nil, err
+	}
+	return conn.ReadStatusWithArguments()
+}
+
+// upload starts an upload.
+func (a *SSHAdapter) upload(t *Transfer, conn *ssh.PktlineConnection, cb ProgressCallback) error {
+	rel, err := t.Rel("upload")
+	if err != nil {
+		return err
+	}
+	if rel == nil {
+		return errors.Errorf("No upload action for object: %s", t.Oid)
+	}
+
+	f, err := os.OpenFile(t.Path, os.O_RDONLY, 0644)
+	if err != nil {
+		return errors.Wrap(err, "basic upload")
+	}
+	defer f.Close()
+
+	status, _, text, err := a.doUpload(t, conn, f, cb)
+	if err != nil {
+		return err
+	}
+	if status < 200 || status > 299 {
+		// A status code of 403 likely means that an authentication token for the
+		// upload has expired. This can be safely retried.
+		if status == 403 {
+			err = errors.New("http: received status 403")
+			return errors.NewRetriableError(err)
+		}
+
+		if status == 429 {
+			return errors.NewRetriableError(fmt.Errorf("got status %d when uploading OID %s", status, t.Oid))
+		}
+
+		if len(text) > 0 {
+			return fmt.Errorf("got status %d when uploading OID %s: %s", status, t.Oid, text[0])
+		}
+		return fmt.Errorf("got status %d when uploading OID %s", status, t.Oid)
+
+	}
+
+	return a.verifyUpload(t, conn)
+}
+
+func (a *SSHAdapter) argumentsForTransfer(t *Transfer, action string) []string {
+	args := make([]string, 0, 3)
+	set, ok := t.Actions[action]
+	if !ok {
+		return nil
+	}
+	args = append(args, fmt.Sprintf("size=%d", t.Size))
+	if set.Id != "" {
+		args = append(args, fmt.Sprintf("id=%s", set.Id))
+	}
+	if set.Token != "" {
+		args = append(args, fmt.Sprintf("token=%s", set.Token))
+	}
+	return args
+}
+
+// Begin a new batch of uploads or downloads. Call this first, followed by
+// one or more Add calls. maxConcurrency controls the number of transfers
+// that may be done at once. The passed in callback will receive updates on
```

Comment:
This comment is presumably a [duplicate](https://github.com/git-lfs/git-lfs/blob/38ae601d5db0da9f94ee3e822d9bd09d4c39da7b/tq/transfer.go#L237-L238) of the one in `tq/transfer.go` for the `Adapter` interface.  Unfortunately I think that comment has been out-of-sync with the actual `Begin()` function's definition since #1774 in 2016, when `maxConcurrency int` was replaced with `cfg AdapterConfig`.  So I guess it would be good to fix both comments now (or just drop this one and fix or ignore the legacy one).

Also, I suspect it may no longer be true that "Either argument may be nil if not required by the client" and in fact at least `cfg` must be defined (but I didn't check this).

Reply:
I've updated the comment.